src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -239,25 +231,21 @@ private void dumpAllTimesToFile(int[][] arrivalTimesAtStopsPerIteration, int max
     }
 
     /**
-     * Before routing, filter the set of patterns down to only the ones that are actually running on the search date.
-     * We can also filter down to only those modes enabled in the search request, because all trips in a pattern are
-     * defined to be on same route, and GTFS allows only one mode per route.
+     * Before routing, filter the set of patterns and trips to only the ones relevant for the search request (date
+     * and transit modes). We set the filtered patterns and trips on the transitLayer, in effect caching them for
+     * repeated searches on the same date with the same modes.
      */
-    private void prefilterPatterns () {
-        for (int patternIndex = 0; patternIndex < transit.tripPatterns.size(); patternIndex++) {
-            TripPattern pattern = transit.tripPatterns.get(patternIndex);
-            RouteInfo routeInfo = transit.routes.get(pattern.routeIndex);
-            TransitModes mode = TransitLayer.getTransitModes(routeInfo.route_type);
-            if (pattern.servicesActive.intersects(servicesActive) && request.transitModes.contains(mode)) {
-                // At least one trip on this pattern is relevant, based on the profile request's date and modes.
-                if (pattern.hasFrequencies) {
-                    runningFrequencyPatterns.set(patternIndex);
-                }
-                // Schedule case is not an "else" clause because we support patterns with both frequency and schedule.
-                if (pattern.hasSchedules) {
-                    runningScheduledPatterns.set(patternIndex);
-                }
-            }
+    private void filterPatternsAndTripsIfNeeded() {
```

Comment:
Would this method be more appropriately part of the TransitLayer? 

`transit.filterPatternsAndTripsIfNeeded(request.transitModes, servicesActive);`

Reply:
It does make sense to put the method on the TransitLayer since it's modifying data within that class. However, the TransitLayer is shared between all threads, so must either be treated as read-only or any modifications must be synchronized. This modification of a single set of filtered patterns is stateful, and the test-then-filter action here, as well as their subsequent use in routing, appears to create a race condition. Specifically, another thread can change the set of filtered patterns while they're still being used, or even while or before they're filtered.

It's not necessarily a problem to concurrently read and write a single reference field serving as a cache, because reference reads are atomic (Java prevents "read tearing" on references). But the reading thread would need to hold on to its own copy of the filtered patterns for a consistent view.